### ISSUE TITLE
Enforce validation of return type of observer methods

### ISSUE THREAD

**Issue by manovotn** [jakartaee/cdi#636]
Currently, CDI doesn't mandate that OM have certain return type.
However, since those methods are invoked by CDI, user has no means to obtain whatever return type/value it has.
It might be worth considering to enforce that the return type is `void` through validation.

Note that in theory, CDI doesn't prohibit users from invoking these methods manually. Due to this, it would be potentially breaking change as some application might be using that. Though all parameters (except event param OFC) of OM are injection points so manual invocation IMO makes little sense.

The only impact of this issue for CDI would be enforcing good practices in CDI applications.

**Comment by Ladicek** [comment]
+1 from me. I actually expected to find a provision in the CDI specification saying that observer methods must be declared `void` and was surprised that I didn't -- even though all observer examples in the specification, as well as all actual observers I've ever seen in practice, _are_ declared `void`.

**Comment by starksm64** [comment]
I'm sure it is just a non-critical detail that was not addressed. Maybe it was seen as too self evident to be worth checking.

**Comment by ljnelson** [comment]
I vote -1. You currently can make almost any existing method you like an observer method (provided it takes parameters) from “outside” via portable extensions. That seems useful.

The same thing is true of initializers (or whatever they’re called; I don’t have the spec handy and I’m on my phone) if I remember right, i.e. `@Inject private Integer setFrob() { return someValue; }`. Something like that.

Historically I don’t think it was an oversight, but a desire to allow maximum tweakability of an existing API from “outside”, via extensions etc. Introducing a breaking change to reduce flexibility is weird. My two cents.

**Comment by manovotn** [comment]
> I vote -1. You currently can make almost any existing method you like an observer method (provided it takes parameters) from “outside” via portable extensions. That seems useful.

Thanks for a viewpoint!

However, do note that initializers are slightly different in that the spec explicitly says that you can directly invoke such method yourself. It doesn't say anything of that sort about observers.
Also, you cannot turn almost any method into observer - sure, you can assume one parameter is the event, but then all others are considered to be injectable beans which might, more often than not, not be the case. I am not sure how useful is this option overall (meaning I have no hard data on it).

**Comment by ljnelson** [comment]
{shrug} I mean, if there's a method in my class like this:
```
public Frob bozo(Gorp gorp) {
  return frob();
}
```
...and my portable extension makes it an observer by adding a synthetic `@Observes` annotation on the parameter, I don't see why I can't have existing code in my application somewhere that calls `frob = bozo(someGorpBuiltByHand);` (as well as obviously the container calling it when a `Gorp` is "fired").  That's just Java.  I don't know why you would want to remove that capability unless there's some deeper reason not mentioned so far in this discussion.

Anyway, the history of the specification early on suggested that Pete and company wanted to preserve existing application semantics as much as possible.  This would seem to be another case where that is true.  Seems odd to remove this.  I continue to vote -1. :-)

**Comment by Ladicek** [comment]
The specific context we had in mind is programmers declaring asynchronous observers that return `CompletionStage<Void>` or similar asynchronous types. To a lot of users, that gives the impression that the CDI container will wait for the completion stage to complete before returning from `fireAsync`, which is not the case. My intuitive reaction to that was: we should treat non-`void` observer methods as definition errors (and I was surprised that the spec doesn't mandate that).

We discussed this issue on the CDI call today and didn't really reach any conclusion AFAICT. I'm tempted to propose that we just close this, citing [Hyrum's Law](https://www.hyrumslaw.com/) as the main reason :-)

**Comment by JHahnHRO** [comment]
An opposing view-point: I'm currently writing a library that integrates RabbitMQ with CDI by firing appropriate Events when messages come in from the broker. One feature I've envisioned is that an interceptor can grab return values of Observer methods and use them for response messages to the broker. I admit that this is a slight abuse of observer methods, but I found it gave very clean looking code, e.g.

```
@RpcEndpoint
Bar handleRabbitMqMessage(@Observes @WithRoutingKey("foo.bar.baz") Foo fooMessage){
//...
}
```

`RpcEndpoint` is an interceptor binding and a corresponding interceptor implementation uses the return value and hands it of to the publisher class. Of course, I could just not do that and insist that users of my library call `Publisher#send` themselves. But if it can be handled automatically, why not?
